File: src/main/java/discretemath/structure/function/AbstractFunction.java
package discretemath.structure.function;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;

import discretemath.structure.exception.DomainIsInfiniteException;
import discretemath.structure.exception.InputNotFromDomainException;
import discretemath.structure.exception.NotInvertibleException;
import discretemath.structure.set.Set;
import discretemath.structure.set.SimpleSet;
import discretemath.structure.set.multi.MultiSet;
import discretemath.structure.set.multi.SimpleMultiSet;
import discretemath.structure.tuple.Pair;
import discretemath.structure.tuple.SimplePair;
import discretemath.utils.CacheValue;

public abstract class AbstractFunction<T, R> implements Function<T, R> {

	protected final CacheValue<Boolean> ontoCache = new CacheValue<>();
	protected final CacheValue<Boolean> oneToOneCache = new CacheValue<>();

	@Override
	public final R apply(T input) throws InputNotFromDomainException, IllegalStateException {
		if (!domain().contains(input)) {
			throw new InputNotFromDomainException();
		}

		return applyWithoutCheckInput(input);
	}

	@Override
	public Set<R> range() {
		if (domain().isInfinite()) {
			throw new DomainIsInfiniteException();
		}

		java.util.Set<R> resultSet = new HashSet<>();

		for (T item : domain()) {
			R result = applyWithoutCheckInput(item);
			resultSet.add(result);
		}

		return SimpleSet.from(resultSet);
	}

	@Override
	public MultiSet<R> multiSetRange() {
		if (domain().isInfinite()) {
			throw new DomainIsInfiniteException();
		}

		Map<R, Integer> rangeSet = new HashMap<>();

		for (T item : domain()) {
			R result = applyWithoutCheckInput(item);

			rangeSet.merge(result, 1, (oldValue, newValue) -> oldValue + 1);
		}

		return SimpleMultiSet.fromCountMap(rangeSet);
	}

	@Override
	public Set<Pair<T, R>> graph() {
		if (domain().isInfinite()) {
			throw new DomainIsInfiniteException();
		}

		java.util.Set<Pair<T, R>> resultSet = new HashSet<>();

		for (T item : domain()) {
			R result = applyWithoutCheckInput(item);
			resultSet.add(new SimplePair<>(item, result));
		}

		return SimpleSet.from(resultSet);
	}

	@Override
	public boolean isOnto() {
		synchronized (ontoCache) {
			if (ontoCache.isSet()) {
				return ontoCache.get();
			}

			boolean isOnto = range().size() == codomain().size();

			ontoCache.set(isOnto);

			return isOnto;
		}
	}

	@Override
	public boolean isOneToOne() {
		synchronized (oneToOneCache) {
			if (oneToOneCache.isSet()) {
				return oneToOneCache.get();
			}

			MultiSet<R> range = multiSetRange();
			R mode = range.mode();
			int modeCount = range.itemCount(mode);

			boolean isOneToOne = modeCount == 1; // no duplicates

			oneToOneCache.set(isOneToOne);

			return isOneToOne;
		}
	}

	@Override
	public Function<R, T> getInverse() throws NotInvertibleException {
		if (!isOnto() || !isOneToOne()) {
			throw new NotInvertibleException();
		}

		if (domain().isInfinite()) {
			throw new DomainIsInfiniteException();
		}

		Map<R, T> reverseMapping = new HashMap<>();

		for (T input : domain()) {
			R result = applyWithoutCheckInput(input);

			reverseMapping.put(result, input);
		}

		return StaticFunction.from(reverseMapping);
	}

	protected abstract R applyWithoutCheckInput(T input);
}

File: src/main/scala/discretemath/structure/function/AbstractFunction.scala
package discretemath.structure.function

import discretemath.structure.exception.{DomainIsInfiniteException, InputNotFromDomainException, NotInvertibleException}
import discretemath.structure.set.main._
import discretemath.structure.set.multi.{MultiSet, SimpleMultiSet}
import discretemath.utils.CacheValue


abstract class AbstractFunction[T, R] extends Function[T, R] {
  protected val ontoCache = new CacheValue[Boolean]
  protected val oneToOneCache = new CacheValue[Boolean]

  @throws[InputNotFromDomainException]
  @throws[IllegalStateException]
  final override def apply(input: T): R = {
    if (!domain.contains(input)) {
      throw new InputNotFromDomainException
    }

    applyWithoutCheckInput(input)
  }

  override def range: Set[R] = {
    if (domain.isInfinite) {
      throw new DomainIsInfiniteException
    }

    SimpleSet(domain.view.map(applyWithoutCheckInput).toSet)
  }

  override def multiSetRange: MultiSet[R] = {
    if (domain.isInfinite) {
      throw new DomainIsInfiniteException
    }

    val rangeSet: Map[R, Int] = domain.view.map(applyWithoutCheckInput).groupMapReduce(identity)(_ => 1)((v1, v2) => v1 + v2)

    SimpleMultiSet(rangeSet)
  }

  override def graph: Set[(T, R)] = {
    if (domain.isInfinite) {
      throw new DomainIsInfiniteException
    }

    val graph = domain.view.map(input => (input, applyWithoutCheckInput(input))).toSet

    SimpleSet(graph)
  }

  override def isOnto: Boolean = {
    ontoCache synchronized {
      if (ontoCache.isSet) return ontoCache.get

      val isOnto = range.size == codomain.size

      ontoCache.set(isOnto)

      return isOnto
    }

  }

  override def isOneToOne: Boolean = {
    oneToOneCache synchronized {
      if (oneToOneCache.isSet) return oneToOneCache.get

      val range = multiSetRange
      val mode: Option[R] = range.mode

      val isOneToOne = mode match {
        case Some(value) => range.itemCount(value) == 1 // no duplicates
        case None => true
      }

      oneToOneCache.set(isOneToOne)

      return isOneToOne
    }

  }

  @throws[NotInvertibleException]
  override def getInverse: Function[R, T] = {
    if (!isOnto || !isOneToOne) throw new NotInvertibleException
    if (domain.isInfinite) throw new DomainIsInfiniteException

    val reverseMapping = domain.view.map(input => (applyWithoutCheckInput(input), input)).toMap

    StaticFunction(reverseMapping)
  }

  protected def applyWithoutCheckInput(input: T): R
}